nsxt/resource_nsxt_upgrade_precheck_acknowledge.go

```diff
@@ -0,0 +1,96 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package nsxt
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	nsxModel "github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/upgrade"
+)
+
+func resourceNsxtUpgradePrecheckAcknowledge() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtUpgradePrecheckAcknowledgeCreate,
+		Read:   resourceNsxtUpgradePrecheckAcknowledgeRead,
+		Update: resourceNsxtUpgradePrecheckAcknowledgeUpdate,
+		Delete: resourceNsxtUpgradePrecheckAcknowledgeDelete,
+
+		Schema: map[string]*schema.Schema{
+			"precheck_ids": {
+				Type:        schema.TypeList,
+				Description: "IDs of precheck warnings that need to be acknowledged",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceNsxtUpgradePrecheckAcknowledgeCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	d.SetId(id)
+	precheckIDs := interface2StringList(d.Get("precheck_ids").([]interface{}))
+	err := acknowledgePrecheckWarnings(m, precheckIDs)
+	if err != nil {
+		return handleCreateError("NsxtPrecheckAcknowledge", id, err)
+	}
+	return resourceNsxtUpgradePrecheckAcknowledgeRead(d, m)
+}
+
+func acknowledgePrecheckWarnings(m interface{}, precheckIDs []string) error {
+	connector := getPolicyConnector(m)
+	client := upgrade.NewPreUpgradeChecksClient(connector)
+	for _, precheckID := range precheckIDs {
+		err := client.Acknowledge(precheckID)
+		if err != nil {
+			return fmt.Errorf("Failed to acknowledge precheck warning with ID %s: %s", precheckID, err)
+		}
+	}
+	return nil
+}
+
+func resourceNsxtUpgradePrecheckAcknowledgeRead(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	precheckWarnings, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_WARNING)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradePrecheckAcknowledge", id, err)
+	}
+	err = setAcknowledgedPrecheckIDsInSchema(d, precheckWarnings)
+	if err != nil {
+		return handleReadError(d, "NsxtUpgradePrecheckAcknowledge", id, err)
+	}
+	return nil
+}
+
+func setAcknowledgedPrecheckIDsInSchema(d *schema.ResourceData, precheckWarnings []nsxModel.UpgradeCheckFailure) error {
+	var precheckWarningIDs []string
+	for _, precheckWarning := range precheckWarnings {
+		if !(*precheckWarning.NeedsAck) {
```

Comment:
Does `NeedsAck` reflect ack state? I understood that this flag indicates whether ack is needed to proceed, not whether it has been acknowledged by the ack API. If my understanding is correct, this would prevent the check from being assigned to state and will cause non-empty state.

Reply:
I am still thinking about what would be the final state of this resource, I think we should have a computed attribute for this resource to show users all prechecks that were acknowledged, or maybe just all precheck failures like the upgrade_prepare resource?